lib/hotel.rb

```diff
@@ -0,0 +1,183 @@
+      end
+    end
+
+    def available_rooms_by_date(year:, month:, day:)
+      valid_date_entry?(year, month, day)
+      available_rooms = []
+      @rooms.each do |room|
+        status = room_reserved?(room_number: room.room_number, year: year, month: month, day: day)
+        if status == false
+          available_rooms << room
+        end
+      end
+      if available_rooms.length == 0
+        puts "There are no available rooms for that date."
+        return nil
+      else
+        return available_rooms 
+      end 
+    end
+
+    def create_block(start_year:, start_month:, start_day:, num_nights:, room_nums:, block_rate:)
+      valid_date_entry?(start_year, start_month, start_day)
+      if room_nums.length > 5
+        raise ArgumentError, "A block can only contain a maximum of 5 rooms."
+      end
+      block_dates = create_date_range(start_year: start_year, start_month: start_month, start_day: start_day, num_nights: num_nights)
+      if room_nums.class != Array
+        raise ArgumentError, "Please enter an array of room numbers for room_nums, even if it has only one room number."
+      end
+      block_rooms = room_nums.map { |num| find_room(num)}
+      block_rooms.each do |room|
+        room.reservations.each do |reservation|
+          overlap = reservation.date_range.overlap?(block_dates)
+          if overlap == true
+            raise NotImplementedError, "Block can't be created, room #{room.room_number} is already booked during those dates."
+          end
+        end
+        room.block_date_ranges.each do |block_date_ranges| 
+          overlap = block_date_ranges.overlap?(block_dates)
+          if overlap == true
+            raise NotImplementedError, "Block can't be created, room #{room.room_number} is already in a block during those dates."
+          end
+        end
+        room.add_block_date_ranges(block_dates)
+      end
+      block_id = create_block_id
+      block = HotelSystem::HotelBlock.new(id: block_id, date_range: block_dates, rooms: block_rooms, room_rate: block_rate)
+      @blocks << block
+    end
+
+    def create_block_id
+      if @blocks.length == nil
+        return 1
+      else
+        id = @blocks.length + 1
+        return id
+      end
+    end
+
+    def find_block(num)
+      return @blocks.find { |block| block.id == num }
+    end
+
+    def reserve_block_room(room_num:, block_id:)
+      room = find_room(room_num)
+      block = find_block(block_id)
+      if block.available_rooms.include?(room)
+        res_id = create_reservation_id
+        block_res = HotelSystem::Reservation.new(id: res_id, room: room, room_number: room_num, date_range: block.date_range, nightly_rate: block.room_rate)
+        @reservations << block_res
+        room.reservations << block_res
+        block.available_rooms.delete(room)
+      else
+        raise NotImplementedError, "Sorry, that room is not available or is not a part of that block."
+      end
+    end
```

Comment:
`NotImplementedError` is not quite the right choice here - that exception is pretty specific in what in means (this method has yet to be implemented by the engineer, or should be implemented in a subclass).

Instead, you should define a custom exception type and raise it here.